src/cs/Ssh.Tcp/Services/StreamForwarder.cs

```diff
@@ -0,0 +1,214 @@
+using System.Threading.Tasks;
+
+namespace Microsoft.DevTunnels.Ssh.Tcp;
+
+public class StreamForwarder : IDisposable
+{
+	private bool disposed;
+
+	public StreamForwarder(
+		Stream localStream,
+		Stream remoteStream,
+		TraceSource trace)
+	{
+		LocalStream = localStream ?? throw new ArgumentNullException(nameof(localStream));
+		RemoteStream = remoteStream ?? throw new ArgumentNullException(nameof(remoteStream));
+		Trace = trace ?? throw new ArgumentNullException(nameof(trace));
+
+		ForwardStreamAsync();
+	}
+
+	public Stream LocalStream { get; }
+
+	public Stream RemoteStream { get; }
+
+	public TraceSource Trace { get; }
+
+	public event EventHandler<EventArgs>? Closed;
+
+	private async void ForwardStreamAsync()
+	{
+		try
+		{
+			bool endOfStream = await ForwardStreamAsync(LocalStream, RemoteStream)
+				.ConfigureAwait(false);
+			Close(abort: !endOfStream);
+		}
+		catch (Exception ex)
+		{
+			// Catch all exceptions in this async void method.
+			Trace.TraceEvent(
+				TraceEventType.Error,
+				SshTraceEventIds.UnknownError,
+				$"{nameof(PortForwardingService)} unexpected error forwarding stream: {ex}");
+			Close(abort: true);
+		}
+	}
+
+	/// <summary>
+	/// Forwards data between local and remote streams until either the end of one of the streams
+	/// is reached or an I/O error occurs.
+	/// </summary>
+	/// <returns>True if the end of stream was reached (graceful closure) else false</returns>
+	protected virtual async Task<bool> ForwardStreamAsync(Stream localStream, Stream remoteStream)
+	{
+		if (localStream == null) throw new ArgumentNullException(nameof(localStream));
+		if (remoteStream == null) throw new ArgumentNullException(nameof(remoteStream));
+
+		using var cancellationSource = new CancellationTokenSource();
+
+		var forwardTask = ForwardStreamAsync(localStream, remoteStream, cancellationSource.Token);
+		var backwardTask = ForwardStreamAsync(remoteStream, localStream, cancellationSource.Token);
+
+		var firstToComplete = await Task.WhenAny(forwardTask, backwardTask)
+			.ConfigureAwait(false);
+		cancellationSource.Cancel();
+
+		await Task.WhenAll(forwardTask, backwardTask).ConfigureAwait(false);
+		return await firstToComplete.ConfigureAwait(false);
+	}
+
+	/// <summary>
+	/// Forwards from a source stream to a destination stream until either the end of the source
+	/// stream is reached, an I/O error occurs, or the cancellation token is cancelled.
+	/// </summary>
+	/// <returns>True if the end of stream was reached (graceful closure) else false</returns>
+	private async Task<bool> ForwardStreamAsync(
+		Stream source,
+		Stream destination,
+		CancellationToken cancellation = default)
+	{
+		byte[] buffer = new byte[8192];
+#if SSH_ENABLE_SPAN
+		var memory = new Memory<byte>(buffer);
+#endif
+		while (!cancellation.IsCancellationRequested)
+		{
+			int count = 0;
+			Exception? readException = null;
+			try
+			{
+#if SSH_ENABLE_SPAN
+				count = await source.ReadAsync(memory, cancellation)
```

Comment:
We could speed up things a bit if we do reading and writing in parallel after the 1st read:
1. read block 1
2. (in parallel) write block 1 and read block 2
3. repeat (2) until read to the end or exception
4. write the last block and flush

Reply:
- That would require 2 buffers, and would be more complicated code. I hesitate to prematurely optimize.
 - We don't do that anywhere else we're forwarding bytes.
 - The incoming stream is buffered internally already, so I doubt that starting the read into another buffer a little sooner will make a significant difference.
 - Normally the remote side of the connection will be much slower than the local side anyway, so forwarding speed will be limited in that way.